rel/value_tuple_dict_entry.go

```diff
@@ -0,0 +1,211 @@
+	return newTuple(
+		NewAttr("@", t.at),
+		NewAttr(DictValueAttr, t.value),
+	)
+}
+
+// Hash computes a hash for a CharTuple.
+func (t DictEntryTuple) Hash(seed uintptr) uintptr {
+	return t.value.Hash(t.at.Hash(seed))
+}
+
+// Equal tests two Tuples for equality. Any other type returns false.
+func (t DictEntryTuple) Equal(v interface{}) bool {
+	if u, ok := v.(DictEntryTuple); ok {
+		return t == u
+	}
+	return false
+}
+
+// String returns a string representation of a Tuple.
+func (t DictEntryTuple) String() string {
+	return fmt.Sprintf("(@: %d, %s: %v)", t.at, DictValueAttr, t.value)
+}
+
+// Eval returns the tuple.
+func (t DictEntryTuple) Eval(local Scope) (Value, error) {
+	return t, nil
+}
+
+var dictValueTupleKind = registerKind(303, reflect.TypeOf((*DictEntryTuple)(nil)))
+
+// Kind returns a number that is unique for each major kind of Value.
+func (t DictEntryTuple) Kind() int {
+	return dictValueTupleKind
+}
+
+// Bool returns true iff the tuple has attributes.
+func (t DictEntryTuple) IsTrue() bool {
+	return true
+}
+
+// Less returns true iff v is not a number or tuple, or v is a tuple and t
+// precedes v in a lexicographical comparison of their name/value pairs.
+func (t DictEntryTuple) Less(v Value) bool {
+	if t.Kind() != v.Kind() {
+		return t.Kind() < v.Kind()
+	}
+	u := v.(DictEntryTuple)
+	if t.at != u.at {
+		return t.at.Less(u.at)
+	}
+	return t.value.Less(u.value)
+}
+
+func (t DictEntryTuple) Negate() Value {
+	return DictEntryTuple{at: t.at.Negate(), value: t.value.Negate()}
+}
+
+// Export exports a Tuple.
+func (t DictEntryTuple) Export() interface{} {
+	return map[string]interface{}{
+		"@":           t.at.Export(),
+		DictValueAttr: t.value.Export(),
+	}
+}
+
+// Count returns how many attributes are in the Tuple.
+func (t DictEntryTuple) Count() int {
+	return 2
+}
+
+// Get returns the Value associated with a name, and true iff it was found.
+func (t DictEntryTuple) Get(name string) (Value, bool) {
+	switch name {
+	case "@":
+		return t.at, true
+	case DictValueAttr:
+		return t.value, true
+	}
+	return nil, false
+}
+
+// MustGet returns e.Get(name) or panics if an error occurs.
+func (t DictEntryTuple) MustGet(name string) Value {
+	if v, has := t.Get(name); has {
+		return v
+	}
+	panic(fmt.Errorf("%q not found", name))
+}
+
+// With returns a Tuple with all name/Value pairs in t (except the one for the
+// given name, if present) with the addition of the given name/Value pair.
+func (t DictEntryTuple) With(name string, value Value) Tuple {
+	return maybeNewDictEntryTupleFromTuple(t.asGenericTuple().With(name, value))
+}
+
+// Without returns a Tuple with all name/Value pairs in t exception the one of
+// the given name.
+func (t DictEntryTuple) Without(name string) Tuple {
+	if name == "@" || name == DictValueAttr {
+		return t.asGenericTuple().Without(name)
+	}
+	return t
+}
+
+func (t DictEntryTuple) Map(f func(Value) Value) Tuple {
+	return maybeNewDictEntryTupleFromTuple(t.asGenericTuple().Map(f))
```

Comment:
why use the `maybe...`?

Reply:
Again, very well spotted! This was probably a copy-paste mistake. For strings and arrays, it's possible for `Map` to produce a non-conforming tuple. This can't happen with `DictEntryTuple`.